Sources/TuistGenerator/Mappers/LibraryResourcesProjectMapper.swift

```diff
@@ -0,0 +1,85 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+/// A project mapper that adds support for defining resources in  libraries
+public class LibraryResourcesProjectMapper: ProjectMapping {
+    public init() {}
+
+    public func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
+        var sideEffects: [SideEffectDescriptor] = []
+        var targets: [Target] = []
+
+        project.targets.forEach { target in
+            let (mappedTargets, targetSideEffects) = mapTarget(target, project: project)
+            targets.append(contentsOf: mappedTargets)
+            sideEffects.append(contentsOf: targetSideEffects)
+        }
+
+        return (project.with(targets: targets), sideEffects)
+    }
+
+    public func mapTarget(_ target: Target, project: Project) -> ([Target], [SideEffectDescriptor]) {
+        guard target.resources.count != 0 else { return ([target], []) }
+        var targets: [Target] = []
+
+        let bundleName = "\(target.name)Resources"
+        let (filePath, fileDescriptors) = synthesizedFile(bundleName: bundleName, target: target, project: project)
+        var modifiedTarget = target
+
+        if target.product.isStatic {
+            let resourcesTarget = Target(name: bundleName,
+                                         platform: target.platform,
+                                         product: .bundle,
+                                         productName: nil,
+                                         bundleId: "\(target.bundleId).resources",
+                                         resources: [],
+                                         filesGroup: target.filesGroup)
+            modifiedTarget.resources = []
+            modifiedTarget.dependencies.append(.target(name: bundleName))
+            targets.append(resourcesTarget)
+        } else {}
+
+        modifiedTarget.sources.append((path: filePath, compilerFlags: nil))
+        targets.append(modifiedTarget)
+
+        return (targets, fileDescriptors)
+    }
+
+    public func synthesizedFile(bundleName: String, target: Target, project: Project) -> (AbsolutePath, [SideEffectDescriptor]) {
+        let filePath = project.path
+            .appending(component: Constants.DerivedDirectory.name)
+            .appending(component: Constants.DerivedDirectory.sources)
+            .appending(component: "Bundle+\(target.name).swift")
+
+        let content: String
+
+        // The bundle is copied into the final product's bundle
+        if target.product.isStatic {
+            content = """
+            import Foundation
+
+            public extension Bundle {
```

Comment:
To avoid name conflicts, we can reduce the visibility such that it is `internal` so it is only scoped to the current module/library/framework.

SwiftPM auto generates the following code snippet which covers a broader set of cases including resources being hosted in frameworks. We could leverage the technique here too for completeness:

```swift
import class Foundation.Bundle

private class BundleFinder {}

extension Foundation.Bundle {
    /// Returns the resource bundle associated with the current Swift module.
    static var module: Bundle = {
        let bundleName = "PackageWithResources_PackageWithResources"

        let candidates = [
            // Bundle should be present here when the package is linked into an App.
            Bundle.main.resourceURL,

            // Bundle should be present here when the package is linked into a framework.
            Bundle(for: BundleFinder.self).resourceURL,

            // For command-line tools.
            Bundle.main.bundleURL,
        ]

        for candidate in candidates {
            let bundlePath = candidate?.appendingPathComponent(bundleName + ".bundle")
            if let bundle = bundlePath.flatMap(Bundle.init(url:)) {
                return bundle
            }
        }
        fatalError("unable to find bundle named PackageWithResources_PackageWithResources")
    }()
}
```

Reply:
Beautiful! Let's take than one then. I was wondering yesterday if we should provide extension for classes like UIImage, or NSImage that have constructors to read the assets from a bundle. What do you think @kwridan ?